// @getmetal/metal-sdk/dist/esm/metal.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import mime from 'mime-types';
import { API_URL, SUPPORTED_FILE_TYPES } from './constants';
import { sanitizeFilename } from './helpers';
import { request } from './request';
export class Metal {
    constructor(apiKey, clientId, indexId) {
        this.apiKey = apiKey;
        this.indexId = indexId;
        this.clientId = clientId;
    }
    index(payload) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const index = (_a = payload.indexId) !== null && _a !== void 0 ? _a : this.indexId;
            if (!index) {
                throw new Error('indexId required');
            }
            const { imageBase64, imageUrl, text, embedding } = payload;
            if (!imageBase64 && !imageUrl && !text && !embedding) {
                throw new Error('payload required');
            }
            const body = { index };
            if (payload === null || payload === void 0 ? void 0 : payload.id) {
                body.id = payload.id;
            }
            if (payload === null || payload === void 0 ? void 0 : payload.metadata) {
                body.metadata = payload.metadata;
            }
            if (imageBase64) {
                body.imageBase64 = imageBase64;
            }
            else if (imageUrl) {
                body.imageUrl = imageUrl;
            }
            else if (text) {
                body.text = text;
            }
            else if (embedding != null) {
                body.embedding = embedding;
            }
            const data = yield request(`${API_URL}/v1/index`, {
                method: 'POST',
                body: JSON.stringify(body),
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    indexMany(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            for (const item of payload) {
                if (!item.index) {
                    item.index = this.indexId;
                }
            }
            const body = { data: payload };
            const data = yield request(`${API_URL}/v1/index/bulk`, {
                method: 'POST',
                body: JSON.stringify(body),
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    search(payload = {}) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const index = (_a = payload.indexId) !== null && _a !== void 0 ? _a : this.indexId;
            if (!index) {
                throw new Error('indexId required');
            }
            const { imageBase64, imageUrl, text, filters } = payload;
            const body = { index, filters };
            if (imageBase64) {
                body.imageBase64 = imageBase64;
            }
            else if (imageUrl) {
                body.imageUrl = imageUrl;
            }
            else if (text) {
                body.text = text;
            }
            const limit = (_b = payload.limit) !== null && _b !== void 0 ? _b : 10;
            let url = `${API_URL}/v1/search?limit=${limit}`;
            if (payload.idsOnly) {
                url += '&idsOnly=true';
            }
            const data = yield request(url, {
                method: 'POST',
                body: JSON.stringify(body),
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    tune(payload) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const index = (_a = payload.indexId) !== null && _a !== void 0 ? _a : this.indexId;
            if (!index) {
                throw new Error('indexId required');
            }
            if (!payload.idA || !payload.idB || Number.isNaN(payload.label)) {
                throw new Error('idA, idB, & label required for payload');
            }
            const body = Object.assign({ index }, payload);
            const data = yield request(`${API_URL}/v1/tune`, {
                method: 'POST',
                body: JSON.stringify(body),
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    getOne(id, indexId) {
        return __awaiter(this, void 0, void 0, function* () {
            const index = indexId !== null && indexId !== void 0 ? indexId : this.indexId;
            if (!id) {
                throw new Error('id required');
            }
            if (!index) {
                throw new Error('indexId required');
            }
            const data = yield request(`${API_URL}/v1/indexes/${index}/documents/${id}`, {
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    getMany(ids, indexId) {
        return __awaiter(this, void 0, void 0, function* () {
            const index = indexId !== null && indexId !== void 0 ? indexId : this.indexId;
            if ((ids === null || ids === void 0 ? void 0 : ids.length) > 100 || (ids === null || ids === void 0 ? void 0 : ids.length) < 1) {
                throw new Error('ids should be between 1 and 100');
            }
            if (!index) {
                throw new Error('indexId required');
            }
            const data = yield request(`${API_URL}/v1/indexes/${index}/documents/${ids.join(',')}`, {
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    deleteOne(id, indexId) {
        return __awaiter(this, void 0, void 0, function* () {
            const index = indexId !== null && indexId !== void 0 ? indexId : this.indexId;
            if (!id) {
                throw new Error('id required');
            }
            if (!index) {
                throw new Error('indexId required');
            }
            const data = yield request(`${API_URL}/v1/indexes/${index}/documents/${id}`, {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    deleteMany(ids, indexId) {
        return __awaiter(this, void 0, void 0, function* () {
            const index = indexId !== null && indexId !== void 0 ? indexId : this.indexId;
            if (!index) {
                throw new Error('indexId required');
            }
            if (!(ids === null || ids === void 0 ? void 0 : ids.length)) {
                throw new Error('ids required');
            }
            const data = yield request(`${API_URL}/v1/indexes/${index}/documents/bulk`, {
                method: 'DELETE',
                body: JSON.stringify({ ids }),
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    createResource(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            const { indexId, fileName, fileType, fileSize } = payload;
            const url = `${API_URL}/v1/indexes/${indexId}/files`;
            const body = {
                fileName: sanitizeFilename(fileName),
                fileType,
            };
            const headers = {
                'Content-Type': 'application/json',
                'x-metal-file-size': fileSize.toString(),
                'x-metal-api-key': this.apiKey,
                'x-metal-client-id': this.clientId,
            };
            const data = yield request(url, {
                method: 'POST',
                body: JSON.stringify(body),
                headers,
            });
            return data;
        });
    }
    uploadFileToUrl(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            const { url, file, fileType, fileSize } = payload;
            const headers = {
                'content-type': fileType,
                'content-length': fileSize.toString(),
            };
            const data = yield request(url, {
                method: 'PUT',
                body: file,
                headers,
            });
            return data;
        });
    }
    uploadFile(file, indexId) {
        return __awaiter(this, void 0, void 0, function* () {
            const index = indexId !== null && indexId !== void 0 ? indexId : this.indexId;
            if (!index) {
                throw new Error('indexId required');
            }
            let fileType;
            let fileSize;
            let fileName;
            let fileData;
            if (typeof file === 'string') {
                const fs = yield import('fs');
                const path = yield import('path');
                fileType = mime.lookup(file) || '';
                fileSize = fs.statSync(file).size;
                fileName = path.basename(file);
                fileData = fs.readFileSync(file);
            }
            else {
                fileType = file.type || '';
                fileSize = file.size;
                fileName = file.name;
                fileData = file;
            }
            if (!SUPPORTED_FILE_TYPES.includes(fileType)) {
                throw new Error('Invalid file type. Supported types are: pdf, docx, csv.');
            }
            const resource = yield this.createResource({ indexId: index, fileName, fileType, fileSize });
            return yield this.uploadFileToUrl({ url: resource.url, file: fileData, fileType, fileSize });
        });
    }
    addDatasource(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!payload.name) {
                throw new Error('name required');
            }
            if (!payload.sourcetype) {
                throw new Error('sourcetype required');
            }
            const url = `${API_URL}/v1/datasources`;
            const data = yield request(url, {
                method: 'POST',
                body: JSON.stringify(payload),
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    getDatasource(id) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!id) {
                throw new Error('id required');
            }
            const url = `${API_URL}/v1/datasources/${id}`;
            const data = yield request(url, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    getAllDatasources(limit, page) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = new URLSearchParams();
            if (limit)
                params.append('limit', limit.toString());
            if (page)
                params.append('page', page.toString());
            const url = `${API_URL}/v1/datasources?${params.toString()}`;
            const data = yield request(url, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    deleteDatasource(id) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!id) {
                throw new Error('id required');
            }
            const url = `${API_URL}/v1/datasources/${id}`;
            const response = yield fetch(url, {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            if (!response.ok) {
                throw new Error(`Error deleting data source: ${response.statusText}`);
            }
        });
    }
    updateDatasource(id, payload) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!id) {
                throw new Error('id required');
            }
            const url = `${API_URL}/v1/datasources/${id}`;
            const data = yield request(url, {
                method: 'PUT',
                body: JSON.stringify(payload),
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    getDataEntity(id) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!id) {
                throw new Error('id required');
            }
            const url = `${API_URL}/v1/data-entities/${id}`;
            const data = yield request(url, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    deleteDataEntity(id) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!id) {
                throw new Error('id required');
            }
            const url = `${API_URL}/v1/data-entities/${id}`;
            const response = yield fetch(url, {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            if (!response.ok) {
                throw new Error(`Error deleting data entity: ${response.statusText}`);
            }
        });
    }
    getAllDataEntities(datasourceId, limit, page) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!datasourceId) {
                throw new Error('id required');
            }
            const params = new URLSearchParams();
            if (limit !== undefined)
                params.append('limit', limit.toString());
            if (page !== undefined)
                params.append('page', page.toString());
            const url = `${API_URL}/v1/datasources/${datasourceId}/data-entities?${params.toString()}`;
            const data = yield request(url, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
    addDataEntityResource(datasource, fileName, fileSize, metadata) {
        return __awaiter(this, void 0, void 0, function* () {
            const url = `${API_URL}/v1/data-entities`;
            const body = {
                datasource,
                name: sanitizeFilename(fileName),
                sourceType: 'file',
                metadata,
            };
            const headers = {
                'Content-Type': 'application/json',
                'x-metal-file-size': fileSize.toString(),
                'x-metal-api-key': this.apiKey,
                'x-metal-client-id': this.clientId,
            };
            return yield request(url, {
                method: 'POST',
                body: JSON.stringify(body),
                headers,
            });
        });
    }
    addDataEntity(payload) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const { datasource, filepath, metadata } = payload;
            if (!datasource) {
                throw new Error("Payload must contain a 'datasource' id");
            }
            let fileType;
            let fileSize;
            let fileName;
            let fileData;
            if (typeof filepath === 'string') {
                const fs = yield import('fs');
                const path = yield import('path');
                fileType = mime.lookup(filepath) || '';
                fileSize = fs.statSync(filepath).size;
                fileName = path.basename(filepath);
                fileData = fs.readFileSync(filepath);
            }
            else {
                const fileObject = filepath;
                fileType = (_a = fileObject.type) !== null && _a !== void 0 ? _a : '';
                fileSize = fileObject.size;
                fileName = fileObject.name;
                fileData = fileObject;
            }
            const dataEntityResource = yield this.addDataEntityResource(datasource, fileName, fileSize, metadata);
            if (!(dataEntityResource === null || dataEntityResource === void 0 ? void 0 : dataEntityResource.url)) {
                throw new Error('Failed to create a data entity resource.');
            }
            // Upload the file to the returned URL
            return yield this.uploadFileToUrl({
                url: dataEntityResource.url,
                file: fileData,
                fileType,
                fileSize,
            });
        });
    }
    addIndex(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!payload.name) {
                throw new Error('Index name is required in payload');
            }
            const body = Object.assign({}, payload);
            const data = yield request(`${API_URL}/v1/indexes`, {
                method: 'POST',
                body: JSON.stringify(body),
                headers: {
                    'Content-Type': 'application/json',
                    'x-metal-api-key': this.apiKey,
                    'x-metal-client-id': this.clientId,
                },
            });
            return data;
        });
    }
}
export default Metal;
